[Mobile] Gallery - Media editing

geriux opened the pull request:
`Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/2496
`WordPress iOS PR` -> https://github.com/wordpress-mobile/WordPress-iOS/pull/14498
`WordPress Android PR` -> https://github.com/wordpress-mobile/WordPress-Android/pull/12497

## Description

Continuing adding media editing support to blocks this PR adds it to Gallery block. ~There aren't any major changes other than adding the `MediaUpload` component in `gallery-image.native` to support replacing images within a gallery~, other changes are wrapping some parts of the code within new containers to fix some absolute positioning issues.

It adds two new props to the `Image` component: `height` and `resizeMode`.

Allows disabling the `Replace` option in the media editing button.

**Note:** _The replace option was removed and will be added in a future iteration._

## How has this been tested?

<details><summary>Steps to test</summary>
---

**Edit an image**

- Open the app with metro running
- Add a Gallery block
- Tap on the block and upload some pictures
- **Expect** to see all the images being uploaded with their progress bar
- Tap on a picture
- **Expect** the image to be selected (blue border around) and the media button option to be visible at the top right
- Tap on the media editing button
- **Expect** to see the options
- Tap on **Edit**
- **Expect** to see the media editor
- Edit the image and tap on done
- **Expect** to see the edited picture uploaded

**Replace an image**

- Open the app with metro running
- Add a Gallery block
- Tap on the block and upload some pictures
- **Expect** to see all the images being uploaded with their progress bar
- Tap on a picture
- **Expect** the image to be selected (blue border around) and the media button option to be visible at the top right
- Tap on the media editing button
- **Expect** to see the options
- Tap on **Replace**
- **Expect** to see the media picker
- Upload a new picture
- **Expect** to see the new picture uploaded and replacing the previous one

**Full screen preview**

- Open the app with metro running
- Add a Gallery block
- Tap on the block and upload some pictures
- **Expect** to see all the images being uploaded with their progress bar
- Tap on a picture
- **Expect** the image to be selected (blue border around) and the media button option to be visible at the top right
- Tap on the image again
- **Expect** to see the image in full screen

**Remove a picture**

- Open the app with metro running
- Add a Gallery block
- Tap on the block and upload some pictures
- **Expect** to see all the images being uploaded with their progress bar
- Tap on a picture
- **Expect** the image to be selected (blue border around) and the media button option to be visible at the top right
- Tap on the media editing button
- **Expect** to see the options
- Tap on **Remove**
- **Expect** to see the selected picture removed

</details>

## Screenshots 

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/88039649-26f69700-cb48-11ea-8fe8-e164104c1a0d.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/88039657-28c05a80-cb48-11ea-8842-d952ede6a600.gif" width="250" /></kbd>

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Commit:
Mobile - Gallery - Media editing

Commit:
Mobile - Update changelog

Commit:
Mobile - Gallery - Remove extra separator

geriux commented:
Hey @iamthomasbishop 👋 now it is turn for Gallery and media editing support. Builds are ready for you to test for a design review whenever you get a chance, thank you!

mkevins reviewed (changes requested):
Excited to see progress on this Gerardo!

I tested this on Android (Pixel 3a - real device) and I think the styles refactoring may have broken a few things regarding scrollable captions:

<img src="https://user-images.githubusercontent.com/8507675/88516488-b1e7fd80-d030-11ea-9af6-acaeeae63d95.gif" width="360">

The caption is no longer scrollable, and the buttons get covered by the large caption. Also, I encountered some issues after the caption expanded beyond the size of the container, where even after deleting the text, the caption remained large, and in one case, weirdly, I couldn't tap out to unfocus the text input. I'm not sure what could cause that behavior. :thinking: 

Another issue I encountered was in using the replace option. Because we are in Gallery, naturally the multiple-selection flag is set, but in this case, using replace, the additional images are added as image blocks _after_ the Gallery block. I'd expect either that the replace option only permits selection of 1 image, or that additionally selected images are inserted into the Gallery block in the position of the image they are replacing.

<details>
<summary>Replace allows multiple, but appends as Image blocks</summary>

Replace allows multiple selection | Multiple images appended as Image blocks
-|-
![gallery-edit-media-button-replace-allows-multiple](https://user-images.githubusercontent.com/8507675/88516942-6f72f080-d031-11ea-9756-05f9d992fd7d.png)|![gallery-edit-media-button-replace-multiple-append-as-image-blocks](https://user-images.githubusercontent.com/8507675/88516954-73067780-d031-11ea-8017-6668ee13517f.png)
</details>

Other than that, all the other behavior seems to be working correctly :+1:

iamthomasbishop commented:
@geriux I had a chance to take the new build for a spin and didn't find any big design issues. I agree w/ @mkevins feedback (nice catches, Matt!), so once those things are done, I think we're in good shape.

One thing that's semi-related but doesn't necessarily have to be part of this work -- can we utilize the mock "iOS system materials" (using the third-party library @geriux used on Compact Notices recently) on the gallery image caption background? Fine by me if we need to roll this into a separate ticket, but Matt's feedback on the caption made me think of this 😄

Commit:
Merge branch 'master' into rnmobile/gallery-media-editing

# Conflicts:
#	packages/react-native-editor/CHANGELOG.md

Commit:
Mobile - Media Editing - Allow disabling the replace option

Commit:
Mobile - Gallery block - Disable replace option for media editing and fix caption styles and scrolling

Commit:
Merge branch 'master' into rnmobile/gallery-media-editing

Commit:
Mobile - Gallery block - Caption styles improvements

Commit:
Mobile - Gallery - Styles improvements for caption

Commit:
Mobile - Gallery - Remove option to replace image

Commit:
Mobile - Gallery - Update styles to use border variable

geriux commented:
Hey @mkevins 👋

Thank you for the review and tests! Nice catches indeed 🙇‍♂️ 

I updated the code and made new builds with the caption fixes, I also removed the `Replace` functionality so we can work on a better approach for Gallery block in another iteration.

Can you please give it another review when you can? Thanks a lot!

geriux commented:
Hey @iamthomasbishop 👋 

> One thing that's semi-related but doesn't necessarily have to be part of this work -- can we utilize the mock "iOS system materials" (using the third-party library @geriux used on Compact Notices recently) on the gallery image caption background? Fine by me if we need to roll this into a separate ticket, but Matt's feedback on the caption made me think of this 😄

That'd be a nice thing to add yes! But I think it'd be better to do it on a separate task. Is there one opened? Thanks!

iamthomasbishop commented:
> That'd be a nice thing to add yes! But I think it'd be better to do it on a separate task. Is there one opened? Thanks!

Most definitely can be a separate task. I will create a ticket and drop some notes on it sometime early next week.

geriux commented:
Hey @lukewalczak 👋 I was wondering if you would be able to review this PR =) 

It's a small one and most of the testing was done already, let me know! Thanks!

lukewalczak commented on packages/block-library/src/gallery/gallery-image.native.js:
```diff
@@ -213,17 +208,31 @@ class GalleryImage extends Component {
+					isSelected={ isSelected }
+					isUploadFailed={ isUploadFailed }
+					isUploadInProgress={ isUploadInProgress }
+					mediaPickerOptions={ [
+						{
+							destructiveButton: true,
+							id: 'removeImage',
+							label: __( 'Remove' ),
+							onPress: onRemove,
+							separated: true,
+							value: 'removeImage',
+						},
+					] }
```
Could you please assign it to the const?

lukewalczak commented on packages/block-library/src/gallery/gallery-image-style.native.scss:
```diff
@@ -1,11 +1,12 @@
-$overlay-border-width: 3px;
+$overlay-border-width: 2px;
 $caption-background-color: rgba(0, 0, 0, 0.4);
 $caption-padding-vertical: 8px;
```
Can we use `$grid-unit` which is `8px` instead of creating `$caption-padding-vertical: 8px`?

lukewalczak commented on packages/block-library/src/gallery/gallery-image.native.js:
```diff
@@ -237,89 +246,71 @@ class GalleryImage extends Component {
+									aria-disabled={ isFirstItem }
+									disabled={ ! isSelected }
+								/>
+								<View style={ separatorStyle }></View>
```
```suggestion
								<View style={ separatorStyle } />
```

Commit:
Mobile - Gallery - Use grid unit for styles and code improvements

Commit:
Merge branch 'master' into rnmobile/gallery-media-editing

# Conflicts:
#	packages/react-native-editor/CHANGELOG.md

geriux commented:
Hey @lukewalczak thank you for your comments! I've just updated the PR with the changes, let me know! Thanks!

lukewalczak reviewed (approved):
LGTM from code perspective!

pinarol commented:
@mkevins is on a leave but his comments are adressed, we completely removed the replace option from this PR. @hypest could you help us merging the PR? Thanks.

hypest commented:
Sure, merging using admin-rights now...
